dtfs-central-api/src/services/state-machines/utilisation-report/event-handlers/helpers/calculate-initial-utilisation-and-fixed-fee.ts

```diff
@@ -0,0 +1,69 @@
+import { calculateInitialUtilisation, isValidDate } from '@ukef/dtfs2-common';
+import { TfmFacilitiesRepo } from '../../../../../repositories/tfm-facilities-repo';
+import { calculateFixedFee } from './calculate-fixed-fee';
+
+export type RequiredParams = {
+  value?: number | null;
+  interestPercentage?: number | null;
+  dayCountBasis?: number | null;
+  coverStartDate?: string | Date | null;
+  coverEndDate?: string | Date | null;
+};
+
+/**
+ * checks that all required values are present
+ * checks that cover start and end dates are in date format
+ * @param value
+ * @param interestPercentage
+ * @param dayCountBasis
+ * @param coverStartDate
+ * @param coverEndDate
+ * @returns true if all values are present and in the correct format
+ */
+export const hasRequiredValues = ({ value, interestPercentage, dayCountBasis, coverStartDate, coverEndDate }: RequiredParams): boolean =>
+  Boolean(
+    value &&
+      coverStartDate &&
+      coverEndDate &&
+      interestPercentage &&
+      dayCountBasis &&
+      isValidDate(new Date(coverStartDate)) &&
+      isValidDate(new Date(coverEndDate)),
+  );
+
+/**
+ * calculateInitialUtilisationAndFixedFee
+ * calculates initial utilisation and fixed fee for a facility
+ * gets the facility from the tfm facilities repo
+ * calculates initial utilisation and fixed fee from facility values
+ * returns the calculated values
+ * @param {String} facilityId
+ * @returns {Object} fixedFee and utilisation values
+ */
+export const calculateInitialUtilisationAndFixedFee = async (facilityId: string) => {
+  const tfmFacility = await TfmFacilitiesRepo.findOneByUkefFacilityId(facilityId);
+
+  if (!tfmFacility?.facilitySnapshot) {
+    throw new Error(`TFM facility ${facilityId} could not be found`);
+  }
+
+  const { value, coverStartDate, coverEndDate, interestPercentage, dayCountBasis } = tfmFacility.facilitySnapshot;
+
+  if (!hasRequiredValues({ value, interestPercentage, dayCountBasis, coverStartDate, coverEndDate })) {
+    throw new Error(`TFM facility values for ${facilityId} are missing`);
+  }
+
+  const utilisation = calculateInitialUtilisation(value);
+  const fixedFee = calculateFixedFee({
+    utilisation,
+    coverStartDate: coverStartDate as Date,
+    coverEndDate: coverEndDate as Date,
```

Comment:
it's because timestamp from the database could be `string | Date | null`.  Even if we use `convertTimestampToDate`, there is a linting error as it expects `string | number | Date`

Reply:
If it can be `string` or `Date` or `null` you shouldn't do `as Date` since it will likely cause an error to be thrown if it isn't a `Date` down the line since all subsequent code will assume it is a `Date`

Reply:
You could write the hasRequiredFields as a user defined type guard instead perhaps to resolve this in a safer way. I think that would look something like this (haven't tested this so might need some changes):
```
export type ValidatedRequiredParams = {
  value: number;
  interestPercentage: number;
  dayCountBasis: number;
  coverStartDate: string | Date;
  coverEndDate: string | Date;
};

export const hasRequiredValues = ({ value, interestPercentage, dayCountBasis, coverStartDate, coverEndDate }: RequiredParams) is ValidatedRequiredParams =>
  Boolean(
    value &&
      coverStartDate &&
      coverEndDate &&
      interestPercentage &&
      dayCountBasis &&
      isValidDate(new Date(coverStartDate)) &&
      isValidDate(new Date(coverEndDate)),
  );
```

Reply:
It will always be Date given `hasRequiredValues` checks it is a valid date.  However, the linting error will keep occurring if I remove as Date

Reply:
It won't necessarily be a `Date` because `isValidDate(new Date(params.coverStartDate))` just checks that it can be converted into a `Date`. It could very well be a string that _can_ be converted into a `Date`.

Reply:
How about simply:
```suggestion
    coverStartDate: new Date(coverStartDate),
    coverEndDate: new Date(coverEndDate),
```

Reply:
It isn't using `convertTimestampToDate` so we are being slightly inconsistent in how we read these values but it is at least safer as a compromise.